src/drivers/rtc/rtc.zig

```diff
@@ -0,0 +1,227 @@
+    valid_ram: bool,
+};
+
+const rtc_address: u16 = 0x70;
+const rtc_data: u16 = 0x71;
+
+const reg_seconds: u8 = 0x0;
+const reg_minutes: u8 = 0x2;
+const reg_hours: u8 = 0x4;
+const reg_day_of_week: u8 = 0x6;
+const reg_day_of_month: u8 = 0x7;
+const reg_month: u8 = 0x8;
+const reg_year: u8 = 0x9;
+var reg_century: u8 = 0x0;
+const reg_status_a: u8 = 0xA;
+const reg_status_b: u8 = 0xB;
+const reg_status_c: u8 = 0xC;
+const reg_status_d: u8 = 0xD;
+
+const DEFAULT_CENTURY: u8 = 20;
+
+fn read_register(reg: u8) u8 {
+    cpu.outb(rtc_address, reg);
+    return cpu.inb(rtc_data);
+}
+
+fn read_register_a() RegisterA {
+    return @bitCast(read_register(reg_status_a));
+}
+
+fn read_register_b() RegisterB {
+    return @bitCast(read_register(reg_status_b));
+}
+
+fn read_register_c() RegisterC {
+    return @bitCast(read_register(reg_status_c));
+}
+
+fn read_register_d() RegisterD {
+    return @bitCast(read_register(reg_status_d));
+}
+
+fn write_register(reg: u8, value: u8) void {
+    cpu.outb(rtc_address, reg);
+    cpu.outb(rtc_data, value);
+}
+
+fn write_register_a(value: RegisterA) void {
+    write_register(reg_status_a, @bitCast(value));
+}
+
+fn write_register_b(value: RegisterB) void {
+    write_register(reg_status_b, @bitCast(value));
+}
+
+fn write_register_c(value: RegisterC) void {
+    write_register(reg_status_c, @bitCast(value));
+}
+
+fn write_register_d(value: RegisterD) void {
+    write_register(reg_status_d, @bitCast(value));
+}
+
+fn bcd_to_bin(bcd: u8) u8 {
+    return ((bcd >> 4) * 10) + (bcd & 0x0F);
+}
+
+fn bin_to_bcd(bin: u8) u8 {
+    return ((bin / 10) << 4) | (bin % 10);
+}
+
+fn is_binary_coded_decimal() bool {
+    return (read_register_b().data_mode) == .Bcd;
+}
+
+pub fn wait_for_rtc() void {
+    @setRuntimeSafety(true);
+    while (read_register_a().update_in_progress) {}
+}
+
+// Read the current time from the RTC without taking into account the update in progress flag
+pub fn get_time() Datetime {
+    wait_for_rtc();
+    return switch (is_binary_coded_decimal()) {
+        true => .{
+            .seconds = read_register(reg_seconds),
+            .minutes = read_register(reg_minutes),
+            .hours = read_register(reg_hours),
+            .day = read_register(reg_day_of_month),
+            .month = read_register(reg_month),
+            .year = read_register(reg_year),
+            .century = if (reg_century != 0) read_register(reg_century) else DEFAULT_CENTURY,
+        },
+        false => .{
+            .seconds = bcd_to_bin(read_register(reg_seconds)),
+            .minutes = bcd_to_bin(read_register(reg_minutes)),
+            .hours = bcd_to_bin(read_register(reg_hours)),
+            .day = bcd_to_bin(read_register(reg_day_of_month)),
+            .month = bcd_to_bin(read_register(reg_month)),
+            .year = bcd_to_bin(read_register(reg_year)),
+            .century = if (reg_century != 0) bcd_to_bin(read_register(reg_century)) else DEFAULT_CENTURY,
+        },
+    };
```

Comment:
I think you may have inverted the condition
```suggestion
    return switch (is_binary_coded_decimal()) {
        false => .{
            .seconds = read_register(reg_seconds),
            .minutes = read_register(reg_minutes),
            .hours = read_register(reg_hours),
            .day = read_register(reg_day_of_month),
            .month = read_register(reg_month),
            .year = read_register(reg_year),
            .century = if (reg_century != 0) read_register(reg_century) else DEFAULT_CENTURY,
        },
        true => .{
            .seconds = bcd_to_bin(read_register(reg_seconds)),
            .minutes = bcd_to_bin(read_register(reg_minutes)),
            .hours = bcd_to_bin(read_register(reg_hours)),
            .day = bcd_to_bin(read_register(reg_day_of_month)),
            .month = bcd_to_bin(read_register(reg_month)),
            .year = bcd_to_bin(read_register(reg_year)),
            .century = if (reg_century != 0) bcd_to_bin(read_register(reg_century)) else DEFAULT_CENTURY,
        },
    };
```

Reply:
In fact, I also inverted the DataMode values.
0 correspond to bcd, and 1 correspond to binary.
That's why it worked even tho I inverted condition here I guess :thinking:

Don't blame me I spent like 10 minutes writing this driver :eyes: